### DISCUSSION TITLE
cpp/memory/cpp_memory_partition

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#66] [Announcements]
# cpp/memory/cpp_memory_partition

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/cpp_memory_partition.html

**Comment by huangjiawei128** [comment]
在 C++ 中，声明 char *c = "abc"; 涉及到两个不同的内存区域：

字符串文字 "abc" 的存储位置：
只读数据段（Read-Only Data Segment）：字符串文字 "abc" 通常存储在只读数据段（.rodata segment）中。这一部分的内存是只读的，确保字符串常量在程序运行时不会被修改。
指针 c 的存储位置：
栈区（Stack）：指针 c 是一个局部变量，存储在栈区。它包含了对只读数据段中字符串 "abc" 的内存地址的引用。

详细说明
只读数据段：包含了程序中所有的字符串文字常量、const 修饰的数据等。它的目的是防止这些常量在运行时被意外修改。
栈区：包含了局部变量，包括指针。char *c 是一个局部变量，它在栈上分配内存，并且指向只读数据段中的 "abc"。

不可修改的内存
尽管 c 指向字符串 "abc"，试图通过 c 修改这个字符串会导致未定义行为。具体来说，尝试写入只读数据段通常会导致程序崩溃，操作系统会检测到非法内存写入并终止程序（例如，产生 segmentation fault）。
